web/src/drafts.js

```diff
@@ -274,6 +275,62 @@ export function update_draft(opts = {}) {
 
 export const DRAFT_LIFETIME = 30;
 
+export function get_draft_ids_for_topic(stream_id, topic) {
+    const drafts = draft_model.get();
+    return Object.keys(drafts)
+        .filter(
+            (draft_id) =>
+                drafts[draft_id].type === "stream" &&
+                drafts[draft_id].stream_id === stream_id &&
+                drafts[draft_id].topic === topic,
+        )
+        .sort((draft_a, draft_b) => drafts[draft_a].updatedAt - drafts[draft_b].updatedAt);
+}
+
+export function get_draft_ids_for_private_message_recipient(private_message_recipient) {
+    const drafts = draft_model.get();
+    return Object.keys(drafts)
+        .filter(
+            (draft_id) =>
+                drafts[draft_id].type === "private" &&
+                _.isEqual(
+                    drafts[draft_id].private_message_recipient.split(",").sort(),
+                    private_message_recipient.split(",").sort(),
+                ),
+        )
+        .sort((draft_a, draft_b) => drafts[draft_a].updatedAt - drafts[draft_b].updatedAt);
+}
+
+export function restore_last_draft_based_on_compose_state() {
+    const message_type = compose_state.get_message_type();
+    if (!message_type || compose_state.message_content()) {
+        return;
+    }
+
+    let last_draft_id;
+    if (message_type === "private") {
+        const recipient = compose_state.private_message_recipient();
+        if (recipient) {
+            last_draft_id = get_draft_ids_for_private_message_recipient(recipient).pop();
+        }
+    } else {
+        const stream_id = compose_state.stream_id();
+        const topic = compose_state.topic();
+        if (stream_id !== undefined && topic !== "") {
+            last_draft_id = get_draft_ids_for_topic(stream_id, topic).pop();
+        }
+    }
```

Comment:
What's the motivation for having these two functions, rather than just a `get_last_drop_id_for_compose_state`?

Actually never mind, we have `filter_drafts_by_compose_box_and_recipient`, which is used in the drafts overlay and does this exact same calculation, but without the bugs noted above. So I propose just moving that function to `drafts.js`, and calling it to find the drafts to consider.

Reply:
Great catch, thanks. Switching to that